core/execution/future/mark_price.go

```diff
@@ -86,38 +94,101 @@ func NewCompositePriceCalculatorFromSnapshot(mp *num.Uint, mpc *snapshot.Composi
+			oracle, err := products.NewCompositePriceOracle(ctx, oe, s, config.SpecBindingForCompositePrice[i], calc.GetUpdateOraclePriceFunc(i))
+			if err != nil {
+				return nil
+			}
+			oracles = append(oracles, oracle)
+		}
+		calc.oracles = oracles
+	}
+	return calc
 }
 
-func NewCompositePriceCalculator(config *types.CompositePriceConfiguration) *CompositePriceCalculator {
+func NewCompositePriceCalculator(ctx context.Context, config *types.CompositePriceConfiguration, oe products.OracleEngine, timeService excommon.TimeService) *CompositePriceCalculator {
 	priceSourcesLen := len(config.SourceStalenessTolerance)
 	if priceSourcesLen == 0 {
 		priceSourcesLen = 1
 	}
 
-	return &CompositePriceCalculator{
+	mpc := &CompositePriceCalculator{
 		config:           config,
 		priceSources:     make([]*num.Uint, priceSourcesLen),
 		sourceLastUpdate: make([]int64, priceSourcesLen),
 		bookPriceAtTime:  map[int64]*num.Uint{},
+		timeService:      timeService,
+	}
+	if len(config.DataSources) > 0 {
+		oracles := make([]*products.CompositePriceOracle, 0, len(config.DataSources))
+		for i, s := range config.DataSources {
+			oracle, err := products.NewCompositePriceOracle(ctx, oe, s, config.SpecBindingForCompositePrice[i], mpc.GetUpdateOraclePriceFunc(i))
+			if err != nil {
+				return nil
+			}
+			oracles = append(oracles, oracle)
+		}
+		mpc.oracles = oracles
 	}
+	return mpc
 }
 
-func (mpc *CompositePriceCalculator) UpdateConfig(config *types.CompositePriceConfiguration) {
+func (mpc *CompositePriceCalculator) UpdateConfig(ctx context.Context, oe excommon.OracleEngine, config *types.CompositePriceConfiguration) error {
+	// special case for only resetting the oracles
+	if config == nil && mpc.oracles != nil {
+		for _, cpo := range mpc.oracles {
+			cpo.UnsubAll(ctx)
+		}
+		return nil
+	}
+
 	priceSourcesLen := len(config.SourceStalenessTolerance)
 	if priceSourcesLen == 0 {
 		priceSourcesLen = 1
 	}
 	mpc.config = config
 	mpc.priceSources = make([]*num.Uint, priceSourcesLen)
 	mpc.sourceLastUpdate = make([]int64, priceSourcesLen)
+	for _, cpo := range mpc.oracles {
+		cpo.UnsubAll(ctx)
+	}
+	mpc.oracles = nil
+	if len(config.DataSources) > 0 {
+		oracles := make([]*products.CompositePriceOracle, 0, len(config.DataSources))
+		for i, s := range config.DataSources {
+			oracle, err := products.NewCompositePriceOracle(ctx, oe, s, config.SpecBindingForCompositePrice[i], mpc.GetUpdateOraclePriceFunc(i))
+			if err != nil {
+				return err
+			}
+			oracles = append(oracles, oracle)
+		}
+		mpc.oracles = oracles
+	}
+	return nil
+}
+
+func (mpc *CompositePriceCalculator) setOraclePriceScalingFunc(f func(context.Context, *num.Numeric) *num.Uint) {
+	mpc.scalingFunc = f
+}
+
+// overridePrice is called to set the price externally. This is used when leaving the opening auction if the
+// methodology yielded no valid price.
+func (mpc *CompositePriceCalculator) overridePrice(p *num.Uint) {
+	if p != nil {
+		mpc.price = p.Clone()
+	}
```

Comment:
How p can be nil? Should we panic here instead to fix it down the line instead of ignoring somethig calling the method with nil?

Reply:
it can be nil because the unit testing framework just says leave auction, not really by uncrossing an auction but forcefully leaving the auction. In real life it can't happen.

Reply:
I agree but not in this PR :)